File: python-server/src/aerospikeworkshop/models/product.py
"""Product model — mirrors com.aerospikeworkshop.model.Product."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class Product(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str | None = None
    brand_name: str | None = Field(default=None, alias="brandName")
    images: dict[str, Any] | None = None
    sub_category: str | None = Field(default=None, alias="subCategory")
    gender: str | None = None
    sale_price: int | None = Field(default=None, alias="salePrice")
    added: int | None = None
    usage: str | None = None
    display_cat: list[str] | None = Field(default=None, alias="displayCat")
    age_group: str | None = Field(default=None, alias="ageGroup")
    colors: list[str] | None = None
    descriptors: dict[str, Any] | None = None
    article_type: str | None = Field(default=None, alias="articleType")
    price: int | None = None
    name: str | None = None
    options: list[dict[str, Any]] | None = None
    season: str | None = None
    article_attr: dict[str, str] | None = Field(default=None, alias="articleAttr")
    variant_name: str | None = Field(default=None, alias="variantName")
    category: str | None = None

    @classmethod
    def from_bins(cls, bins: dict[str, Any]) -> Product:
        return cls.model_validate(bins)

    def to_bins(self) -> dict[str, Any]:
        return self.model_dump(by_alias=True, exclude_none=False)

File: rust-server/src/models/product.rs
use std::collections::HashMap;

use aerospike::Value;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

use super::{model_to_bins, record_to_model};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Product {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub brand_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub images: Option<JsonValue>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sub_category: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gender: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sale_price: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub added: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub usage: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_cat: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub age_group: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub colors: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub descriptors: Option<JsonValue>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub article_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<Vec<JsonValue>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub season: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub article_attr: Option<HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variant_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
}

impl Product {
    pub fn from_bins(bins: &HashMap<String, Value>) -> Result<Self, serde_json::Error> {
        record_to_model(bins)
    }

    pub fn to_bins(&self) -> HashMap<String, Value> {
        model_to_bins(self)
    }
}
